vpn/tls.go

```diff
@@ -28,9 +28,66 @@ var (
 	ErrBadKeypair = errors.New("bad keypair conf")
 	// ErrBadParrot is returned for errors during TLS parroting
 	ErrBadParrot = errors.New("cannot parrot")
+	// ErrCannotVerifyCertChain is returned for certificate chain validation errors.
+	ErrCannotVerifyCertChain = errors.New("cannot verify chain")
 )
 
-// initTLS returns a tls.Config matching the VPN options.
+// certVerifyOptionsNoCommonNameCheck returns a x509.VerifyOptions initialized with
+// an empty string for the DNSName. This allows to skip CN verification.
+func certVerifyOptionsNoCommonNameCheck() x509.VerifyOptions {
+	return x509.VerifyOptions{DNSName: ""}
+}
+
+// certVerifyOptions is the option that the customVerify function will use; by
+// default is the version that skips the DNSName check.
+var certVerifyOptions = certVerifyOptionsNoCommonNameCheck
+
+// customVerify is a version of the verification routines that does not try to verify
+// the Common Name, since we don't know it a priori for a VPN gateway. Returns
+// an error if the verification fails.
+// From tls/common documentation: If normal verification is disabled by
+// setting InsecureSkipVerify, [...] then this callback will be considered but
+// the verifiedChains argument will always be nil.
+func customVerify(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
+	roots := x509.NewCertPool()
+	var leaf *x509.Certificate
+
+	for i, rawCert := range rawCerts {
+		cert, _ := x509.ParseCertificate(rawCert)
+
+		if cert != nil {
+			// we assume (from docs) that we're always given the
+			// leaf certificate as the first cert in the array.
+			if i == 0 {
+				leaf = cert
+			} else {
+				roots.AddCert(cert)
+			}
+		}
+	}
+
+	if roots == nil {
+		return fmt.Errorf("%w: %s", ErrCannotVerifyCertChain, "roots is nil")
+	}
```

Comment:
Are we testing for this case? I'm a bit confused here because I see that `roots` is initialized with `NewCertPool` and that function [does not seem to return a nil value](https://cs.opensource.google/go/go/+/refs/tags/go1.18.3:src/crypto/x509/cert_pool.go;l=59)

Reply:
Perhaps a different algorithm? Something like:

```
if len(rawCerts) < 2 {
  return errors.New("missing a cert and roots")
}
// fill the left and the pool
```

Also, is it possible for us to get `len(rawCerts) == 1`? Could a provider only have a single cert?

If this could happen, then perhaps we could accept code with `len(rawCerts) == 1` and treat the case in which we have more than a single certificate specially?

I see that the full document for the callback says:

```Go
	// VerifyPeerCertificate, if not nil, is called after normal
	// certificate verification by either a TLS client or server. It
	// receives the raw ASN.1 certificates provided by the peer and also
	// any verified chains that normal processing found. If it returns a
	// non-nil error, the handshake is aborted and that error results.
	// [...]
	VerifyPeerCertificate func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error
```

So, I _think_, at least in theory, it's possible for a server to send us a single certificate, right?

Reply:
Also, for documenting our processes here, did we consider using [VerifyConnection](https://pkg.go.dev/crypto/tls#example-Config-VerifyConnection) to verify the connection rather than `VerifyPeerCertificate`?

Why did we chose `VerifyPeerCertificate` over `VerifyConnection`?